dirserver/config.txt
# Configuration Options:
#
# BaseListenPort = ####
#     Default: 4990
#     Description: The port on which the server should listen for client connections
#
# UpstreamServers = dir1.example.com, dir2.example.com:1234, ...
#     Default: sscentral.sscuservers.net, ssdir.playsubspace.com, warpath.bluetoast.org
#     Description: A list of directory servers to mirror
#
UpstreamServers = ssdir.playsubspace.com, ds1.krslynx.com, ds1.hlrse.net, sscentral.sscuservers.net, dirserver.ssnecentral.net
# ZoneRefreshSeconds = ####
#     Default: 60
#     Description: Rate at which zones are fetched from upstream servers
#
# ZoneBanPrefix = prefix1, prefix2, ...
#     Default:
#     Description: Zones with any of these prefixes will be removed from the directory listing
#
# ZoneBanString = string1, string2, ...
#     Default:
#     Description: Zones with names containing any of these strings will be removed from the directory listing
#
# ZoneBanNetmask = cidr_netmask1, cidr_netmask2, ...
#     Default:
#     Description: Zones with IP addresses that are in the subnet specified by the CIDR netmask will be removed from
#                  the directory listing.
#     Example: ZoneBanNetmask = 192.168.0.0/16, 172.16.0.0/12, 10.0.0.0/8
#
# PasswordReclaimSeconds = ####
#     Default: 86400
#     Description: The amount of time before a zone name becomes available again (i.e. password change)
#
# StatisticsWriteSeconds = ####
#     Default: 60
#     Description: The amount of time between two successive dumps of the server's statistics
#
LogFile = /dev/stdout
#     Default:
#     Description: The log file name for the server. Logging is disabled if the filename is empty.
#
LogDebug = 1
#     Default:
#     Description: Enables debugging information in the log if LogDebug = 1. This has to be set after LogFile.
#
# TimeToLive = #
#     Default: 5
#     Description: Specifies the visible zone update subgraph based on this graph diameter.


starlight/CHANGES
????-??-??
Release 1.4

	- fixed bug where non-public arenas starting with a number could not be entered
	- added support for accented (ISO-8859-1) characters (reported by Audry)
	- fixed bug where older subgames didn't route chat messages correctly (reported
	  by froot)

2008-12-28
Release 1.3

	- changed compiler and installer for easier deployment
	- added support for themes
	- arena list uses themed highlight color instead of light blue
	- removed background color option - this should be set via themes
	- remote private messages now displayed as :playerName:message (Continuum-style)
	- messages to chat channel 1 are not sent as ;1;message anymore
	- fixed protocol bug that caused position packets to be ignored
	- fixed occasional map checksum errors in ASSS zones
	- worked around GTK's minimize->restore window bug on Windows

2008-07-26
Release 1.2

	- fixed bug where application would become unresponsive if the internet connection
	  was broken
	- added chat logging (saved to: data/logs/session.log)
	- added a notify list: when a message containing one of the notify strings is
	  received, the icon changes color and the application flashes in the taskbar
	- the zone selection is now maintained after receiving updates from the directory server
	- added a dialog box to display Continuum-style custom login errors
	- added ?ignore command - use ?ignore add <name> and ?ignore remove <name> to add and
	  remove players from the ignore list
	- icon changes color in taskbar and system tray when a private message is received
	  and the window is not focused
	- expands %selfname, %tickname, and %squad
	- added option to not display timestamps
	- added enter/leave messages
	- added player kill messages
	- displays player's squad in player list and allows sorting by name or squad
	- added registration form for new users - the lack of registration form prevented
	  the creation of new accounts on the SSC biller in previous versions (thanks to
	  CRe> for pointing this out)
	- fixed CRe>'s email address in About dialog
	- upgrades to the underlying control protocol (much more efficient - especially
	  on upgraded ASSS servers)

2008-06-06
Release 1.1

	- new Windows installer that correctly sets up shortcuts and icons
	- changed application icon (thanks to CRe> for the artwork)
	- window flashes in taskbar when not focused and a private message received
	  (previously only flashed when minimized)
	- added configuration item to change background color of chat and input
	  areas
	- input text color now always matches public text color
	- chat area auto-scrolls when the scrollbar is completely scrolled
	- added :: functionality to send message to last user who sent a private
	  message
	- added a "minimize to tray" feature

2008-05-30
Release 1.0

    - Initial release


dirserver/src/Config.cpp
/*++

Copyright (c) 2010 Sharvil Nanavati.

Module:

    Config.cpp

Authors:

    Sharvil Nanavati (sharvil) 2010-05-06

--*/

#include "base.h"
#include "Config.h"
#include "LogLib.h"

#include <cstring>

const string Config::KConfigFile("config.txt");

Config::Config()
	: iLogger(Logger::Instance()),
	  iFileLogSink(0)
{
	RestoreDefaults();
}

Config::~Config()
{
	delete iFileLogSink;
}

bool Config::Load()
{
	try
	{
		uint32 fileTime = File::LastModifiedTime(KConfigFile);
		if(fileTime == iLastModifiedTime)
			return false;

		File fp(KConfigFile, "rt");

		RestoreDefaults();
		iLastModifiedTime = fileTime;

		while(!fp.IsEndOfFile())
		{
			string line = fp.ReadLine();
			string::size_type pos = line.find('#');

			if(pos != string::npos)
				line.erase(pos);

			pos = line.find('=');
			if(pos == string::npos)
				continue;

			string key = line.substr(0, pos);
			string value = line.substr(pos + 1);

			if(String::Trim(key).empty() || String::Trim(value).empty())
				continue;

			if(String::Compare(key, "UpstreamServers", String::KCaseInsensitive) == 0)
			{
				iUpstreamServers.clear();

				StringSequence servers = String::Split(value, ',');
				for(StringSequence::iterator i = servers.begin(); i != servers.end(); ++i)
					if(!String::Trim(*i).empty())
					{
						if(i->find(':') == string::npos)
							iUpstreamServers.insert(*i + ":4990");
						else
							iUpstreamServers.insert(*i);
					}
			}
			else if(String::Compare(key, "PasswordReclaimSeconds", String::KCaseInsensitive) == 0)
			{
				uint32 tmp;
				if(sscanf(value.c_str(), "%u", &tmp) == 1)
					iPasswordReclaimSeconds = tmp;
			}
			else if(String::Compare(key, "ZoneRefreshSeconds", String::KCaseInsensitive) == 0)
			{
				uint32 tmp;
				if(sscanf(value.c_str(), "%u", &tmp) == 1)
					iZoneListDownloadSeconds = tmp;
			}
			else if(String::Compare(key, "StatisticsWriteSeconds", String::KCaseInsensitive) == 0)
			{
				uint32 tmp;
				if(sscanf(value.c_str(), "%u", &tmp) == 1)
					iStatisticsDumpSeconds = tmp;
			}
			else if(String::Compare(key, "BaseListenPort", String::KCaseInsensitive) == 0)
			{
				uint32 port;
				if(sscanf(value.c_str(), "%u", &port) == 1 && port < 65536)
					iBaseListenPort = port;
			}
			else if(String::Compare(key, "TimeToLive", String::KCaseInsensitive) == 0)
			{
				uint32 ttl;
				if(sscanf(value.c_str(), "%u", &ttl) == 1)
					iTtl = std::min(32U, ttl);
			}
			else if(String::Compare(key, "ZoneBanPrefix", String::KCaseInsensitive) == 0)
			{
				StringSequence prefixes = String::Split(value, ',');
				for(StringSequence::iterator i = prefixes.begin(); i != prefixes.end(); ++i)
					if(!String::ToLower(String::Trim(*i)).empty())
						iZoneBanPrefixes.insert(*i);
			}
			else if(String::Compare(key, "ZoneBanString", String::KCaseInsensitive) == 0)
			{
				StringSequence strings = String::Split(value, ',');
				for(StringSequence::iterator i = strings.begin(); i != strings.end(); ++i)
					if(!String::ToLower(String::Trim(*i)).empty())
						iZoneBanStrings.insert(*i);
			}
			else if(String::Compare(key, "ZoneBanNetmask", String::KCaseInsensitive) == 0)
			{
				StringSequence netmasks = String::Split(value, ',');
				for(StringSequence::iterator i = netmasks.begin(); i != netmasks.end(); ++i)
					try
					{
						if(!String::Trim(*i).empty())
							iZoneBanNetmasks.push_back(CidrNetmask(*i));
					}
					catch(const std::runtime_error & error)
					{
						iLogger.Log(KLogWarning, "%s", error.what());
					}
			}
			else if(String::Compare(key, "LogFile", String::KCaseInsensitive) == 0)
			{
				try
				{
					iFileLogSink = new FileLogSink(value);
					iLogger.AttachSink(*iFileLogSink);
				}
				catch(const not_found_error &)
				{
					iLogger.Log(KLogWarning, "Unable to open log file %s.\n", value.c_str());
				}
			}
			else if(String::Compare(key, "LogDebug", String::KCaseInsensitive) == 0)
			{
				uint32 debug = 0;
				if(sscanf(value.c_str(), "%u", &debug) == 1 && debug != 0)
					iLogger.SetLogThreshold(KLogDebug);
			}
		}
	}
	catch(const not_found_error & error)
	{
		if(iLastModifiedTime == 0)
			return false;

		RestoreDefaults();
	}
	return true;
}

void Config::RestoreDefaults()
{
	delete iFileLogSink;
	iFileLogSink = 0;
	iLogger.SetLogThreshold(KLogInfo);

	iUpstreamServers.clear();
	iUpstreamServers.insert("sscentral.sscuservers.net:4990");
	iUpstreamServers.insert("ssdir.playsubspace.com:4990");
	iUpstreamServers.insert("warpath.bluetoast.org:4990");

	iPasswordReclaimSeconds = 3600 * 24;
	iZoneListDownloadSeconds = 60;
	iStatisticsDumpSeconds = 60;
	iConfigReloadSeconds = 10;
	iLastModifiedTime = 0;
	iBaseListenPort = 4990;
	iTtl = 5;
	iZoneBanPrefixes.clear();
	iZoneBanStrings.clear();
	iZoneBanNetmasks.clear();
}

const StringISet & Config::GetUpstreamServerList() const
{
	return iUpstreamServers;
}

uint32 Config::GetPasswordReclaimTime() const
{
	return iPasswordReclaimSeconds;
}

uint32 Config::GetZoneListRefreshPeriod() const
{
	return iZoneListDownloadSeconds;
}

uint32 Config::GetStatisticsDumpPeriod() const
{
	return iStatisticsDumpSeconds;
}

uint32 Config::GetConfigReloadPeriod() const
{
	return iConfigReloadSeconds;
}

uint16 Config::GetBaseListenPort() const
{
	return iBaseListenPort;
}

bool Config::IsZoneBanned(const InetAddress & ip, const string & name) const
{
	return IsZoneIpBanned(ip) || IsZoneNameBanned(name);
}

bool Config::IsZoneIpBanned(const InetAddress & ip) const
{
	for(NetmaskList::const_iterator i = iZoneBanNetmasks.begin(); i != iZoneBanNetmasks.end(); ++i)
		if(ip << *i)
			return true;

	return false;
}

bool Config::IsZoneNameBanned(const string & name) const
{
	string canonicalName = String::ToLowerCopy(name);

	for(StringISet::const_iterator i = iZoneBanPrefixes.begin(); i != iZoneBanPrefixes.end(); ++i)
		if(memcmp(canonicalName.c_str(), i->c_str(), std::min(canonicalName.length(), i->length())) == 0)
			return true;

	for(StringISet::const_iterator i = iZoneBanStrings.begin(); i != iZoneBanStrings.end(); ++i)
		if(canonicalName.find(*i) != string::npos)
			return true;

	return false;
}

bool Config::IsAlive(uint32 hopCount) const
{
	return hopCount < iTtl;
}
